src/opustrainer/alignments.py

```diff
@@ -0,0 +1,25 @@
+from typing import Optional, List
+from opustrainer.types import Pair, TokenList
+
+
+def parse_alignments(pairs:str, src_tokens:Optional[TokenList]=None, trg_tokens:Optional[TokenList]=None) -> List[Pair]:
+    pairs = [
+        Pair(int(a), int(b)) for a, b in
+        (pair.split('-', maxsplit=1) for pair in pairs.split())
+    ]
+
+    if src_tokens is not None and trg_tokens is not None:
+        invalid_pairs = [
+            pair
+            for pair in pairs
+            if pair.src < 0 or pair.src >= len(src_tokens)
+            or pair.trg < 0 or pair.trg >= len(trg_tokens)
+        ]
+        if invalid_pairs:
+            raise ValueError('Out-of-bound alignment pairs: ' + ' '.join(map(repr, invalid_pairs)))
```

Comment:
Hmm, I agree. What should we do with lines that have out-of-bound pairs? I think we should just throw out all alignment pairs in that case for that sentence because there's no guarantee that the rest of them are not off-by-one or something.

That would automatically also exclude them from the *Tags* modifier, and their alignment info wouldn't be passed on to Marian, so Marian would not be trained with broken data (or worse, bork because of an index error).

Reply:
I would throw out the lines, if we can do it cleanly. If alignment wasn't produced, this likely means the sentence is bad.

Reply:
Okay, euhm. I can do that, but it has to happen at the modifier-is-applied-to-batch level in trainer.py because that's the only place we can remove lines at the moment. I'll change that.

There's a discussion somewhere with Graeme about changing modifiers to return a list of lines, which then can just add and remove lines. But that's probably too much also add into this pull request.

Reply:
or maybe then merge as is and start another PR aiming at sanitizing input.